Adds a new concept page for autoscaling (#44959)
kubernetes/website: content/en/docs/concepts/workloads/autoscaling.md

---
title: Autoscaling Workloads
description: >-
  With autoscaling, you can automatically update your workloads in one way or another. This allows your cluster to react to changes in resource demand more elastically and efficiently.
content_type: concept
weight: 40
---

<!-- overview -->

In Kubernetes, you can _scale_ a workload depending on the current demand of resources.
This allows your cluster to react to changes in resource demand more elastically and efficiently.

There are manual and automatic ways to scale your workloads, depending on your use case.

<!-- body -->

## Scaling workloads manually

Kubernetes supports _manual scaling_ of workloads, either by changing the number of

> {{< glossary_tooltip text="Replicas" term_id="replica">}} of an object that manages a set of

Reviewer: (nit)
```suggestion
{{< glossary_tooltip text="replicas" term_id="replica">}} define for an object that manages a set of
```

{{< glossary_tooltip text="Pods" term_id="pod" >}} (for example a {{< glossary_tooltip text="Deployment" term_id="deployment" >}}),
or by adjusting the provided resources of each Replica (for example CPU or memory).

> ### Scaling the number of Replicas of a workload

Reviewer: ```suggestion
### Manual changes to replica count
```

As this is a a concept pages, you're defining concepts (like the idea that you can manually change replica count).
Separately, we could make a task page about this - perhaps under https://kubernetes.io/docs/tasks/run-application/ - that reiterates how to run `kubectl scale`.

Author: I need some clarification on this - do you want me to not go into detail regarding how to (manually) scale workloads at all, or do you want me to adjust the section headers to make it sound less of a tutorial, but keep the short paragraphs below?

For these paragraphs, I took a look at the [Deployment concept page](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) which also gives short outlines of e.g. how to create deployments using a manifest.

> You can use the `kubectl scale` command to increase or decrease the number of Replicas of a workload:

Reviewer: ```suggestion
You can use the `kubectl scale` command to increase or decrease the number of replicas for a workload:
```

```shell

> kubectl scale deployment <deployment-name> --replicas=<desired-replicas>

Reviewer: (nit)
```suggestion
kubectl scale deployment <deployment-name> --replicas=<desired-count>
```

```

See also this [example of scaling a Deployment](/docs/concepts/workloads/controllers/deployment/#scaling-a-deployment) in the `Deployment` documentation.

### Resizing workloads in-place

> Instead of scaling the number of Replicas of a workload, you can also adjust the provided resources
> of each Replica, in-place. You do this by patching the entries in one or both of the following
> fields of the `Pod` or [PodTemplate](/docs/concepts/workloads/pods/#pod-templates) you want to resize:

Reviewer: ```suggestion
Instead of scaling the number of replicas of a workload, you can also adjust the provided resources
for a particular pod or set of pods, in-place. You do this by patching the entries in one or both of the following
fields of the `Pod` or [pod template](/docs/concepts/workloads/pods/#pod-templates) you want to resize:
```

- `spec.containers[*].resources.requests`
- `spec.containers[*].resources.limits`

{{< caution >}}

> Resizing a workload in-place **without** restarting the Pods or its Containers requires Kubernetes version 1.27 or later.

Reviewer: Do you, as a cluster administrator, need to ensure that any feature gates are enabled? (I think you do, `InPlacePodVerticalScaling`).

Author: Yup, it's an alpha staged feature flag. How do I hint at that? By adding text to the `note` shortcode or by putting a `feature-state` shortcode in the top of the concept page?

Reviewer: I'd add a subheading, and follow the subheading with a feature state shortcode.

> {{< /caution >}}

Reviewer: Use a `note`. `caution` is for advice you should follow or you put your cluster at risk.

See also this task about [resizing CPU and memory resources](/docs/tasks/configure-pod-container/resize-container-resources) assigned to Containers.

## Scaling workloads automatically

Kubernetes also supports _automatic scaling_ of workloads, which is the focus of this page.

The concept of _Autoscaling_ in Kubernetes refers to the ability to automatically update the workloads of your cluster. This can be either an object that manages a set of Pods (for example a

> {{< glossary_tooltip text="Deployment" term_id="deployment" >}} or Pods or PodTemplates themselves.

Reviewer: AFAIK, you can't autoscale a bare Pod, nor a PodTemplate.

Author: Ah, misleading wording on my side. I meant 'Pods or PodTemplates' as in 'even when scaling deployments vertically, its the PodTemplate's specs within the deployment that are being updated'.
Will improve this part.

Depending on _what_ is being scaled, there are also different options for _how_ to scale:

- scale the number of available instances (such as Replicas in a Deployment)
- scale the available resources on existing instances themselves (such as CPU or memory of a {{< glossary_tooltip text="Container" term_id="container" >}} in a Pod)

> The first option is referred to as _horizontal scaling_, while the second is referred to as _vertical scaling_.

Reviewer: Consider introducing this explanation before the automatic scaling section. Manual horizontal scaling is still horizontal scaling.

I'd consider expanding the page intro to cover this detail, or make it the first heading (eg “Vertical and horizontal scaling”).

Author: Good point, will do. 
I was hesitant to put more stuff in the `overview` part since the page content guidelines tell you to [only use a single paragraph for the overview section](https://kubernetes.io/docs/contribute/style/page-content-types/#concept).

### Scaling workloads horizontally

In Kubernetes, you can automatically scale a workload horizontally using a _HorizontalPodAutoscaler_ (HPA).

It is implemented as a Kubernetes API resource and a {{< glossary_tooltip text="controller" term_id="controller" >}}

> and periodically adjusts the number of {{< glossary_tooltip text="Replicas" term_id="replica" >}}

Reviewer: ```suggestion
and periodically adjusts the number of {{< glossary_tooltip text="replicas" term_id="replica" >}}
```

in a workload to match observed resource utilization such as CPU or memory usage.

> There is a [walkthrough example](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough) of configuring a HorizontalPodAutoscaler for a Deployment.

Reviewer: ```suggestion
There is a [walkthrough tutorial](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough/) of configuring a HorizontalPodAutoscaler for a Deployment.
```

(a separate PR should move that tutorial into the tutorials section; anyway, it _is_ a tutorial).

Author: Right. Do you want me to open a separate issue for someone to do this?

---
title: Autoscaling Workloads
description: >-
  With autoscaling, you can automatically update your workloads in one way or another. This allows your cluster to react to changes in resource demand more elastically and efficiently.
content_type: concept
weight: 40
---

<!-- overview -->

In Kubernetes, you can _scale_ a workload depending on the current demand of resources.
This allows your cluster to react to changes in resource demand more elastically and efficiently.

When you scale a workload, you can either increase or decrease the number of replicas managed by
the workload, or adjust the resources available to the replicas in-place.

The first approach is referred to as _horizontal scaling_, while the second is referred to as
_vertical scaling_.

There are manual and automatic ways to scale your workloads, depending on your use case.

<!-- body -->

## Scaling workloads manually

Kubernetes supports _manual scaling_ of workloads. Horizontal scaling can be done
using the `kubectl` CLI.

> For vertical scaling, you need to _patch_ the resource definition of your workload.

Reviewer 2: should there be a note about VPA ? It's an extra install so I would say no but it is part of k/autoscaler
https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler

Author: If I don't get you wrong, there is a whole paragraph about the VPA:

<img width="706" alt="Screenshot 2024-02-16 at 17 26 51" src="https://github.com/kubernetes/website/assets/28064429/03705bec-340f-4352-9ce1-1aa21785905e">

Or are you referring to something else entirely?

Reviewer: https://k8s.io/docs/concepts/workloads/autoscaling/#scaling-workloads-vertically

---

> reviewers: []

Reviewer: ```suggestion
```

title: Autoscaling Workloads
description: >-

>   With Autoscaling, you can automatically update your workloads or infrastructure in one way or another. This allows your cluster to react to changes in resource demand more elastically and efficiently.

Reviewer: “autoscaling” or “automatic scaling” please.

content_type: concept
weight: 40

> hide_summary: true # Listed separately in section index

Reviewer: ```suggestion
```
unless you edit the section index to make that true

---

<!-- overview -->

The concept of _Autoscaling_ in Kubernetes refers to the ability to automatically update the available
resources of your cluster. This can be either a replication controller (for example a
{{< glossary_tooltip text="Deployment" term_id="deployment" >}} or
{{< glossary_tooltip text="ReplicaSet" term_id="replica-set" >}}), or the cluster infrastructure

> itself (for example the number of {{< glossary_tooltip text="Nodes" term_id="node" >}}).

Reviewer: Avoid using the term “replication controller”; it's likely to mislead. We'd now say something like “an object that lets you manage a set of Pods”.

Besides the differentiation in _what_ is being scaled, there are also different options for _how_ to scale:

- scale the number of available instances (such as Pods or Nodes)
- scale the available resources on existing instances themselves (such as CPU or memory)

The first option is referred to as _horizontal scaling_, while the second is referred to as _vertical scaling_.

<!-- body -->

> 

Reviewer: You might like to discuss manual scaling first: you can either change the scale of an object, or use [in-place resizing](https://kubernetes.io/docs/tasks/configure-pod-container/resize-container-resources/).

Choose whether you want to explain both of those before you talk about any automatic scaling, or if you'd like to split _Scaling workloads horizontally_ into manual and then automatic (and the same for vertical scaling).

> ## Scaling Workloads Horizontally

Reviewer: We write headings (not page titles, but headings) in [sentence case](https://kubernetes.io/docs/contribute/style/style-guide/#headings)

In Kubernetes, you can scale a workload horizontally using a _HorizontalPodAutoscaler_ (HPA).
It is implemented as a Kubernetes API resource and a {{< glossary_tooltip text="controller" term_id="controller" >}}
and periodically adjusts the number of {{< glossary_tooltip text="replicas" term_id="replica" >}}
in a workload to match observed resource utilization such as CPU or memory usage.

> There is a [walkthrough example](../../../tasks/run-application/horizontal-pod-autoscale-walkthrough.md) of configuring a HorizontalPodAutoscaler for a Deployment.

Reviewer 3: Some early review of the draft content, the current way of specifying the link is incorrect and leads to a 404 error.  Consider using suggested way below.

```suggestion
There is a [walkthrough example](/docs/tasks/run-application/horizontal-pod-autoscale-walkthrough) of configuring a HorizontalPodAutoscaler for a Deployment.
```

Additionally, you can also verify how this content renders by checking the build preview link: [deploy-preview/docs/concepts/workloads/autoscaling/](https://deploy-preview-44959--kubernetes-io-main-staging.netlify.app/docs/concepts/workloads/autoscaling/)

## Scaling Workloads Vertically

> _tba_ about VerticalPodAutoscaler

Reviewer: Also mention in-place manual autoscaling.

## Scaling the Cluster

> _tba_ about Cluster Autoscaler and Karpenter

Reviewer: Given the title of “autoscaling workloads”, a discussion about node autoscaling belongs in a different page, I think. It's definitely OK to include a _summary_ of the topic in this page.

> ## Advanced Scenarios

Reviewer: I'd make a heading for each kind of scaling (and avoid the term “advanced”):

- Event driven automatic scaling
- Sizing a workload to match your cluster
- Scheduled scaling
  (this is for things like turning off a workload at weekends)
